fetch.bs

```diff
@@ -8449,6 +8597,60 @@ with a <var>promise</var>, <var>request</var>, <var>responseObject</var>, and an
 </div>
 
 
+<h3 id=fetch-later-method>FetchLater method</h3>
+
+<pre class=idl>
+
+dictionary DeferredRequestInit : RequestInit {
+  DOMHighResTimeStamp backgroundTimeout;
```

Comment:
Per reason discussions with our OT users, it looks like there is also some need for `pageHideTimeout` in addition to `backgroundTimeout`. Should we consider replace this with a general `timeout` which kicks off after calling `fetchLater()`, and let user freely decide whichever event to with it with?

Reply:
what's the difference? Just a general visibility change, like minimizing the window?

Reply:
This is tricky because the page may be frozen after visibilitychange and no JS timers will fire I think there is a footgun here. Browsers may suspend JS timers after visibilitychange. Explicitly, this happens on the `freeze` event but I suspect it might happen at other times too (`freeze` is experimental and [not available in](https://caniuse.com/mdn-api_document_freeze_event) WebKit or Mozilla, I think).

So that makes it maybe impossible reliably implement send-after-hidden (if some browsers stop timers). Even in those that send a `freeze` event before stopping timers, it's complex to get correct, you need to

- set a timer
- listen to freeze
- if either happens, send the beacon immediately and prevent the other event handler from doing so. E.g.

So we may need to give users a way to set a timeout after hidden. Maybe a way to say "timeout after JS timers become unreliable" covers this and BFCache?

Reply:
Yea I think this would be the same background timeout (we can find a different name for it).
We can say in the HTML spec, that if the page frozen due to visibility change, start the inactive timer for the fetch group (this doesn't affect this PR!)

Reply:
@noamr @fergald A followup question to the definition of our fetchGroup's `deactivated` state:

In addition to the `freeze` Page LifeCycle event, should it also include `pause` event or not?

Per https://html.spec.whatwg.org/C/#pause, no loads are allowed to start/continue in this state, and all background processing is also paused: `... require the user agent to pause while running a task until a condition goal is met`

I would assume the answer is no, but would like to confirm.

Reply:
This is unfortunate but it seems like we should not *send* while paused or we risk breaking code like

```
if (!fetchLaterResult.activated) {
  ...
  abortController.abort();
  ...
}
```

So there's not much point in starting a timer on pause. I guess it doesn't matter a whole lot either way. If a page cares about the time the beacon was sent, it will have its own timer running in JS. Either our timer sends it or their timer sends it.

Reply:
Pausing is not related to deactivation. The main thread is paused when doing things like fetching module scripts or `alert()`. As far as we're concerned the page is active during pause.